registry/confdb_control.go

```diff
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+var (
+	validAccountID = regexp.MustCompile("^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28})$")
+)
+
+type AuthenticationMethod string
+
+const (
+	OperatorKey AuthenticationMethod = "operator-key"
+	Store       AuthenticationMethod = "store"
+)
+
+// isValidAuthenticationMethod checks if a string is a valid AuthenticationMethod
+func isValidAuthenticationMethod(value string) bool {
+	switch AuthenticationMethod(value) {
+	case OperatorKey, Store:
+		return true
+	default:
+		return false
+	}
+}
+
+// convertToAuthenticationMethod converts and validates a []string to []AuthenticationMethod
+func convertToAuthenticationMethod(methods []string) ([]AuthenticationMethod, error) {
+	sort.Slice(methods, func(i, j int) bool {
+		return methods[i] < methods[j]
+	})
+
+	// remove duplicates
+	methods = compact(methods)
+
+	var result []AuthenticationMethod
+	for _, method := range methods {
+		if !isValidAuthenticationMethod(method) {
+			return nil, fmt.Errorf("invalid authentication method: %s", method)
+		}
+		result = append(result, AuthenticationMethod(method))
+	}
+	return result, nil
+}
+
+// Operator holds the delegations for a single operator.
+type Operator struct {
+	ID     string
+	Groups []*Group
+}
+
+// Group holds a set of views delegated through the given authentication.
+type Group struct {
```

Comment:
Wondering whether `DelegatedViews` or `ControlGroup` wouldn't be better as Group is fairly generic but it's minor anyway

Reply:
I was trying to match the spec but I agree, something like `ControlGroup` is clearer.

Thanks for the review!

Reply:
given that is shows up as registry.Group I think we definitely need a more precise name, group in the context of the assertion is fine is it is scoped to the assertion, but the type name here is not

Reply:
@pedronis, I've renamed it to `ControlGroup`... `registry.ControlGroup` or `confdb.ControlGroup` seems better.